gluster/subvol/cmd/gluster-subvol-provisioner/gluster-subvol-provisioner.go

```diff
@@ -0,0 +1,608 @@
+	// created. This is the same behaviour as provisioners that do not
+	// provide cloning support. The "k8s.io/CloneOf" annotation should only
+	// get set if cloning was successful.
+	sourcePVCRef, err := p.tryClone(options.PVC, mountpoint, destDir)
+	if err != nil || sourcePVCRef == "" {
+		err = os.MkdirAll(destDir, 0775)
+		if err != nil {
+			return nil, fmt.Errorf("Failed to create subdir for new pvc %s: %s", options.PVC.Name, err)
+		}
+		glog.V(1).Infof("successfully created Gluster Subvol %+v with size and volID", destDir)
+	}
+
+	var gid *int
+	if gidAllocate {
+		var allocate int // suggested by vetshadow, warning for 'allocate, err := ..'
+		allocate, err = p.allocator.AllocateNext(options)
+		if err != nil {
+			return nil, fmt.Errorf("allocator error: %v", err)
+		}
+		gid = &allocate
+	}
+	glog.V(1).Infof("Allocated GID %d for PVC %s", *gid, options.PVC.Name)
+
+	// TODO: set quota? Not possible through standard tools, only gluster CLI?
+
+	// subdir has been setup, create the PVC object
+	ep, err := p.copyEndpoints(supervolNS, supervolPV, options.PVC.Namespace, options.PVC.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to copy endpoint from supervol %s: %s", supervolPV.Name, err)
+	}
+
+	// TODO: glusterfile creates a Service for each PVC, is that really needed?
+
+	glusterfs := &v1.GlusterfsVolumeSource{
+		EndpointsName: ep.ObjectMeta.Name,
+		Path:          p.makeMountPath(supervolPV, options.PVC.Namespace, options.PVC.UID),
+		ReadOnly:      false,
+	}
+
+	mode := v1.PersistentVolumeFilesystem
+	pvSpec := v1.PersistentVolumeSpec{
+		PersistentVolumeReclaimPolicy: options.PersistentVolumeReclaimPolicy,
+		AccessModes:                   options.PVC.Spec.AccessModes,
+		MountOptions:                  supervolPV.Spec.MountOptions,
+		VolumeMode:                    &mode,
+		Capacity: v1.ResourceList{
+			v1.ResourceName(v1.ResourceStorage): options.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)],
+		},
+		PersistentVolumeSource: v1.PersistentVolumeSource{
+			Glusterfs: glusterfs,
+		},
+	}
+
+	pv := &v1.PersistentVolume{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: options.PVName,
+			Annotations: map[string]string{
+				gidAnn:         strconv.FormatInt(int64(*gid), 10),
+				glusterTypeAnn: "subvol",
+				parentPVCAnn:   supervolNS + "/" + supervolPVC.Name,
+				"Description":  descAnn,
+			},
+		},
+		Spec: pvSpec,
+	}
+
+	// sourcePVCRef will be empty if there was no cloning request, or cloning failed
+	if sourcePVCRef != "" {
+		options.PVC.Annotations[CloneOfAnn] = sourcePVCRef
+		err = p.annotatePVC(options.PVC.Namespace, options.PVC.Name, options.PVC.Annotations)
+		if err != nil {
+			// TODO: retry or cleanup the cloned data?
+			glog.Errorf("cloning %s was successful, but setting the annotation on PVC %s/%s was not", sourcePVCRef, options.PVC.Namespace, options.PVC.Name)
+		}
+	}
+
+	return pv, nil
+}
+
+func (p *glusterSubvolProvisioner) Delete(pv *v1.PersistentVolume) error {
```

Comment:
If supervol was not mounted this code does a mount before mounting subvol. I dont see we do the reverse here if no other references exist ? if yes, it could leave an unwanted/stale mount of parent/supervol in the node which is really risky.

Reply:
This provisioner does not mount the subvol, it only creates/deletes the subdirectory for the subvol on the supervol Gluster volume.

There is indeed no unmounting of the supervol. Because the filesystem runs in userspace, the mount process will get killed when this pod exits/restarts. At that time the supervol automatically gets unmounted.

Mounting is a relatively slow operation. It would not be very efficient to mount+unmount for each PVC that gets provisioned/deleted. I guess an enhancement for this provisioner would be to expire and unmount supervols after a certain time of not being used. At the moment I do not see a strong need for it though.

Reply:
>There is indeed no unmounting of the supervol. Because the filesystem runs in userspace, the mount process will get killed when this pod exits/restarts. At that time the supervol automatically gets unmounted.
>

This is my point. There is a stale mount of supervol till the pod exits/crashes which is not good IMO.  The mount presence till somehow there is an ungraceful shutdown of pod looks awkward as well. What about failing the provision() if supervol is not mounted ?

Reply:
The supervol will get mounted in `Provision()`, if not mounted yet. In case mounting fails, an error is reported (see line ~190).